Move implementation of empty_like for sparse COO

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68084
* #68083
* __->__ #71103

Previously the implementation of empty_like for sparse COO was a
conditional path in generic implementation.
This PR makes use of the Dispatcher and moves the implementation into a
separate function.

cc @nikitaved @pearu @cpuhrsch

Differential Revision: [D33511240](https://our.internmc.facebook.com/intern/diff/D33511240)

IvanYashchuk committed:
Move implementation of empty_like for sparse COO

Previously the implementation of empty_like for sparse COO was a
conditional path in generic implementation.
This PR makes use of the Dispatcher and moves the implementation into a
separate function.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Move implementation of empty_like for sparse COO"


Previously the implementation of empty_like for sparse COO was a
conditional path in generic implementation.
This PR makes use of the Dispatcher and moves the implementation into a
separate function.

cc nikitaved pearu cpuhrsch

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33511240).

cpuhrsch commented:
@pytorchbot ciflow rerun

cpuhrsch commented:
@IvanYashchuk - the failures are real.

```
ERROR [0.006s]: test_sparse_zeros_neg_cpu_uint8 (__main__.TestSparseUnaryUfuncsCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 753, in test_wrapper
    return test(*args, **kwargs)
  File "test_sparse.py", line 3487, in test_sparse_zeros
    actual = op(sparse_input)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_methods_invocations.py", line 777, in __call__
    return self.op(*args, **kwargs)
RuntimeError: options.layout() == kSparse && self.is_sparse()INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/aten/src/ATen/native/sparse/SparseTensor.cpp":836, please report a bug to PyTorch. 
```

IvanYashchuk committed:
Update on "Move implementation of empty_like for sparse COO"


Previously the implementation of empty_like for sparse COO was a
conditional path in generic implementation.
This PR makes use of the Dispatcher and moves the implementation into a
separate function.

cc nikitaved pearu cpuhrsch

Differential Revision: [D33511240](https://our.internmc.facebook.com/intern/diff/D33511240)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Move implementation of empty_like for sparse COO"


Previously the implementation of empty_like for sparse COO was a
conditional path in generic implementation.
This PR makes use of the Dispatcher and moves the implementation into a
separate function.

cc nikitaved pearu cpuhrsch

Differential Revision: [D33511240](https://our.internmc.facebook.com/intern/diff/D33511240)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33511240).
